presto-main/src/main/java/com/facebook/presto/sql/planner/plan/DistinctLimitNode.java

```diff
@@ -83,18 +85,16 @@ public boolean isPartial()
         return hashSymbol;
     }
 
+    @JsonProperty
     public List<Symbol> getDistinctSymbols()
     {
-        if (hashSymbol.isPresent()) {
-            return ImmutableList.copyOf(Iterables.filter(getOutputSymbols(), not(hashSymbol.get()::equals)));
-        }
-        return getOutputSymbols();
+        return distinctSymbols;
     }
 
     @Override
     public List<Symbol> getOutputSymbols()
     {
-        return source.getOutputSymbols();
+        return distinctSymbols;
```

Comment:
Hm. Then it should be `source.getOutputSymbols()` (to preserve layout) filtered by `distinctSymbols`. If `distinctSymbols` set is different than `source.getOutputSymbols()` set at the end of planning then validator fails. This would make  `DistinctLimitNode` symbolically correct during intermediate optimizations. Still, we rely on `PruneUnreferencedSymbols` to cut child output symbols, but we don't loose the information which symbols should be distinct. This fixes issues with potentially rewritten source nodes (by join pruning, `MergeProjections` and `MergeWindows`). What do you think @martint ?